### ISSUE TITLE
Second step cannot get correct input value from loop item

### ISSUE THREAD

**Issue by TrevorNodeJS** [argoproj/argo-workflows#1793]
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG 
**What happened**:
call  docker/whalesay:latest cowsay two times in every loop.
second call cannot print out correct message. Print our the item[0]'s content: hello world abc001sap.comabc001.

Because input parameters are item[0].

**What you expected to happen**:
Second step should print out correct message. 

**How to reproduce it (as minimally and precisely as possible)**:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: loops-two-steps-
spec:
  entrypoint: loop-example-main
  arguments:
    parameters:
      - name: tenants
        value: |
          [
            { "tenant-id": "abc001", "subdomain": "sap.com" , "message": "hello world abc001"},
            { "tenant-id": "ibm001", "subdomain": "ibm" , "message": "hello world ibm"},
            { "tenant-id": "apple001", "subdomain": "apple" , "message": "hello world apple"}
          ]

  templates:
  - name: loop-example-main
    parallelism: 1
    inputs:
      parameters:
        - name: tenants
    steps:
    - - name: loop-two-steps
        template: loop-two-steps
        arguments:
          parameters:
            - name: message
              value: "{{item.message}}"
            - name: tenant-id
              value: "{{item.tenant-id}}"
            - name: subdomain
              value: "{{item.subdomain}}"
        withParam: "{{inputs.parameters.tenants}}" 


  - name: loop-two-steps
    inputs:
      parameters:
      - name: message
      - name: tenant-id
      - name: subdomain

    steps:
    - - name: print-message-1
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "{{inputs.parameters.message}}" 
          - name: tenant-id
            value: "{{inputs.parameters.tenant-id}}" 
          - name: subdomain
            value: "{{inputs.parameters.subdomain}}"

    - - name: print-message-2
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "{{inputs.parameters.message}}" 
          - name: tenant-id
            value: "{{inputs.parameters.tenant-id}}"
          - name: subdomain
            value: "{{inputs.parameters.subdomain}}"

  - name: whalesay
    inputs:
      parameters:
      - name: message
      - name: tenant-id
      - name: subdomain
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["{{inputs.parameters.message}}{{inputs.parameters.subdomain}}{{inputs.parameters.tenant-id}}"]

```

**Anything else we need to know?**:

**Environment**:
- Argo version:
```bash
$ argo version
argo: v2.4.2
  BuildDate: 2019-10-21T18:40:37Z
  GitCommit: 675c66267f0c916de0f233d8101aa0646acb46d4
  GitTreeState: clean
  GitTag: v2.4.2
  GoVersion: go1.11.5
  Compiler: gc
  Platform: windows/amd64
```
- Kubernetes version :
```bash
$ kubectl version -o yaml
clientVersion:
  buildDate: "2019-10-15T12:11:03Z"
  compiler: gc
  gitCommit: 211047e9a1922595eaa3a1127ed365e9299a6c23
  gitTreeState: clean
  gitVersion: v1.14.8
  goVersion: go1.12.10
  major: "1"
  minor: "14"
  platform: windows/amd64
serverVersion:
  buildDate: "2019-10-15T12:02:12Z"
  compiler: gc
  gitCommit: 211047e9a1922595eaa3a1127ed365e9299a6c23
  gitTreeState: clean
  gitVersion: v1.14.8
  goVersion: go1.12.10
  major: "1"
  minor: "14"
  platform: linux/amd64
```

**Other debugging information (if applicable)**:
- workflow result:
```
$ argo get <workflowname>

- executor logs:
```
$ kubectl logs <failedpodname> -c init
$ kubectl logs <failedpodname> -c wait
```
- workflow-controller logs:
```
$ kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

**Comment by TrevorNodeJS** [comment]
sorry click wrong button.

**Comment by sarabala1979** [comment]
@TrevorNodeJS 
I am not able to reproduce above issue with given workflow

```
 k logs -n workflows   loops-two-steps-6crhw-399920394 main
 __________________________ 
< hello world ibmibmibm001 >
 -------------------------- 
    \
     \
      \     
                    ##        .            
              ## ## ##       ==            
           ## ## ## ##      ===            
       /""""""""""""""""___/ ===        
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~   
       \______ o          __/            
        \    \        __/             
          \____\______/   
Name:                loops-two-steps-6crhw
Namespace:           workflows
ServiceAccount:      default
Status:              Succeeded
Created:             Sun Nov 24 22:41:31 -0800 (4 minutes ago)
Started:             Sun Nov 24 22:41:31 -0800 (4 minutes ago)
Finished:            Sun Nov 24 22:42:10 -0800 (3 minutes ago)
Duration:            39 seconds
Parameters:          
  tenants:           [
  { "tenant-id": "abc001", "subdomain": "sap.com" , "message": "hello world abc001"},
  { "tenant-id": "ibm001", "subdomain": "ibm" , "message": "hello world ibm"},
  { "tenant-id": "apple001", "subdomain": "apple" , "message": "hello world apple"}
]


STEP                                          PODNAME                           DURATION  MESSAGE
 ✔ loops-two-steps-6crhw (loop-example-main)                                              
 └-·-✔ loop-two-steps(0:) (loop-two-steps)                                                
   | ├---✔ print-message-1 (whalesay)         loops-two-steps-6crhw-3858883007  6s        
   | └---✔ print-message-2 (whalesay)         loops-two-steps-6crhw-3740400051  5s        
   ├-✔ loop-two-steps(1:) (loop-two-steps)                                                
   | ├---✔ print-message-1 (whalesay)         loops-two-steps-6crhw-399920394   5s        
   | └---✔ print-message-2 (whalesay)         loops-two-steps-6crhw-2118744806  5s        
   └-✔ loop-two-steps(2:) (loop-two-steps)                                                
     ├---✔ print-message-1 (whalesay)         loops-two-steps-6crhw-1487433581  5s        
     └---✔ print-message-2 (whalesay)         loops-two-steps-6crhw-2501638565  6s        

$ k logs -n workflows loops-two-steps-6crhw-1487433581  main
 ________________________________ 
< hello world appleappleapple001 >
 -------------------------------- 
    \
     \
      \     
                    ##        .            
              ## ## ##       ==            
           ## ## ## ##      ===            
       /""""""""""""""""___/ ===        
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~   
       \______ o          __/            
        \    \        __/             
          \____\______/

**Comment by TrevorNodeJS** [comment]
@sarabala1979 sarabala1979

could you please check logs for steps: print-message-2 (whalesay)

loops-two-steps-6crhw-3740400051
loops-two-steps-6crhw-2118744806 
loops-two-steps-6crhw-2501638565

then, you will get same value for different branches.

```bash
STEP                                                                                                     PODNAME                           DURATION  MESSAGE
 ✔ loops-two-steps-jl75v (loop-example-main)
 └-·-✔ loop-two-steps(0:message:hello world abc001,subdomain:sap.com,tenant-id:abc001) (loop-two-steps)
   | ├---✔ print-message-1 (whalesay)                                                                    loops-two-steps-jl75v-4125227693  19s
   | └---✔ print-message-2 (whalesay)                                                                    loops-two-steps-jl75v-2300656869  22s
   ├-✔ loop-two-steps(1:message:hello world ibm,subdomain:ibm,tenant-id:ibm001) (loop-two-steps)
   | ├---✔ print-message-1 (whalesay)                                                                    loops-two-steps-jl75v-2354878282  17s
   | └---✔ print-message-2 (whalesay)                                                                    loops-two-steps-jl75v-3248337702  16s
   └-✔ loop-two-steps(2:message:hello world apple,subdomain:apple,tenant-id:apple001) (loop-two-steps)
     ├---✔ print-message-1 (whalesay)                                                                    loops-two-steps-jl75v-1837811903  16s
     └---✔ print-message-2 (whalesay)                                                                    loops-two-steps-jl75v-1719328947  16s

I039142@PVGN50947922A MINGW64 ~ (master)
$ kubectl logs loops-two-steps-jl75v-3248337702  main
 _________________________________
< hello world abc001sap.comabc001 >
 ---------------------------------
    \
     \
      \
                    ##        .
              ## ## ##       ==
           ## ## ## ##      ===
       /""""""""""""""""___/ ===
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
       \______ o          __/
        \    \        __/
          \____\______/

```

**Comment by sarabala1979** [comment]
yes. I can reproduce the issue. Thanks for information. looking for help to fix. can you fix and create PR?

**Comment by niebniebnieb** [comment]
See same issue. Also retries seem to be affected.

**Comment by simster7** [comment]
While looking into this issue I found this bug: https://github.com/argoproj/argo/issues/1799. Could be related

**Comment by simster7** [comment]
This issue is actually fixed by #1744. I built that branch and had these results (I replaced the parameters with `{A, B, C} X {1, 2, 3}` for easier readability; `pl` here just prints the logs):

```
$ argo watch loops-two-steps-sndzt
Name:                loops-two-steps-sndzt
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Wed Nov 27 07:49:22 -0800 (1 minute ago)
Started:             Wed Nov 27 07:49:22 -0800 (1 minute ago)
Finished:            Wed Nov 27 07:49:58 -0800 (50 seconds ago)
Duration:            36 seconds
Parameters:
  tenants:           [
  { "tenant-id": "A1", "subdomain": "A2" , "message": "A3"},
  { "tenant-id": "B1", "subdomain": "B2" , "message": "B3"},
  { "tenant-id": "C1", "subdomain": "C2" , "message": "C3"}
]


STEP                                          PODNAME                           DURATION  MESSAGE
 ✔ loops-two-steps-sndzt (loop-example-main)
 └-·-✔ loop-two-steps(0:) (loop-two-steps)
   | ├---✔ print-message-1 (whalesay)         loops-two-steps-sndzt-3141327238  4s
   | └---✔ print-message-2 (whalesay)         loops-two-steps-sndzt-2252043922  5s
   ├-✔ loop-two-steps(1:) (loop-two-steps)
   | ├---✔ print-message-1 (whalesay)         loops-two-steps-sndzt-2427952395  4s
   | └---✔ print-message-2 (whalesay)         loops-two-steps-sndzt-3656928879  5s
   └-✔ loop-two-steps(2:) (loop-two-steps)
     ├---✔ print-message-1 (whalesay)         loops-two-steps-sndzt-2287009132  4s
     └---✔ print-message-2 (whalesay)         loops-two-steps-sndzt-1272804148  5s

$ pl loops-two-steps-sndzt-2252043922
 ________
< A3A2A1 >
 --------
    \
     \
      \
                    ##        .
              ## ## ##       ==
           ## ## ## ##      ===
       /""""""""""""""""___/ ===
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
       \______ o          __/
        \    \        __/
          \____\______/
$ pl loops-two-steps-sndzt-3656928879
 ________
< B3B2B1 >
 --------
    \
     \
      \
                    ##        .
              ## ## ##       ==
           ## ## ## ##      ===
       /""""""""""""""""___/ ===
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
       \______ o          __/
        \    \        __/
          \____\______/
$ pl loops-two-steps-sndzt-1272804148
 ________
< C3C2C1 >
 --------
    \
     \
      \
                    ##        .
              ## ## ##       ==
           ## ## ## ##      ===
       /""""""""""""""""___/ ===
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
       \______ o          __/
        \    \        __/
          \____\______/
```

**Comment by simster7** [comment]
Hello @TrevorNodeJS, [`v2.4.3`](https://github.com/argoproj/argo/releases/tag/v2.4.3) was just released and should have the fix for this issue, please take a look!
